lxd/storage/drivers/driver_cephfs.go

```diff
@@ -0,0 +1,977 @@
+	if volType != VolumeTypeCustom {
+		return nil, fmt.Errorf("Volume type not supported")
+	}
+
+	snapshotDir, err := GetVolumeSnapshotDir(d.name, volType, volName)
+	if err != nil {
+		return nil, err
+	}
+
+	snapshots := []string{}
+
+	ents, err := ioutil.ReadDir(snapshotDir)
+	if err != nil {
+		// If the snapshots directory doesn't exist, there are no snapshots.
+		if os.IsNotExist(err) {
+			return snapshots, nil
+		}
+
+		return nil, err
+	}
+
+	for _, ent := range ents {
+		fileInfo, err := os.Stat(filepath.Join(snapshotDir, ent.Name()))
+		if err != nil {
+			return nil, err
+		}
+
+		if !fileInfo.IsDir() {
+			continue
+		}
+
+		snapshots = append(snapshots, ent.Name())
+	}
+
+	return snapshots, nil
+}
+
+func (d *cephfs) RestoreVolume(vol Volume, snapshotName string, op *operations.Operation) error {
+	sourcePath := GetVolumeMountPath(d.name, vol.volType, vol.name)
+	cephSnapPath := filepath.Join(sourcePath, ".snap", snapshotName)
+
+	// Restore using rsync.
+	bwlimit := d.config["rsync.bwlimit"]
+	output, err := rsync.LocalCopy(cephSnapPath, vol.MountPath(), bwlimit, false)
+	if err != nil {
+		return fmt.Errorf("Failed to rsync volume: %s: %s", string(output), err)
+	}
+
+	return nil
+}
+
+func (d *cephfs) MigrateVolume(vol Volume, conn io.ReadWriteCloser, volSrcArgs migration.VolumeSourceArgs, op *operations.Operation) error {
+	if vol.volType != VolumeTypeCustom {
+		return fmt.Errorf("Volume type not supported")
+	}
+
+	if vol.contentType != ContentTypeFS {
+		return fmt.Errorf("Content type not supported")
+	}
+
+	if volSrcArgs.MigrationType.FSType != migration.MigrationFSType_RSYNC {
+		return fmt.Errorf("Migration type not supported")
+	}
+
+	bwlimit := d.config["rsync.bwlimit"]
+
+	for _, snapName := range volSrcArgs.Snapshots {
+		snapshot, err := vol.NewSnapshot(snapName)
+		if err != nil {
+			return err
+		}
+
+		// Send snapshot to recipient (ensure local snapshot volume is mounted if needed).
+		err = snapshot.MountTask(func(mountPath string, op *operations.Operation) error {
+			var wrapper *ioprogress.ProgressTracker
+			if volSrcArgs.TrackProgress {
+				wrapper = migration.ProgressTracker(op, "fs_progress", snapshot.name)
+			}
+
+			path := shared.AddSlash(mountPath)
+			return rsync.Send(snapshot.name, path, conn, wrapper, nil, bwlimit, d.state.OS.ExecPath)
+		}, op)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Send volume to recipient (ensure local volume is mounted if needed).
+	return vol.MountTask(func(mountPath string, op *operations.Operation) error {
+		var wrapper *ioprogress.ProgressTracker
+		if volSrcArgs.TrackProgress {
+			wrapper = migration.ProgressTracker(op, "fs_progress", vol.name)
+		}
+
+		path := shared.AddSlash(mountPath)
+		return rsync.Send(vol.name, path, conn, wrapper, nil, bwlimit, d.state.OS.ExecPath)
+	}, op)
+}
+
+func (d *cephfs) CreateVolumeFromMigration(vol Volume, conn io.ReadWriteCloser, volTargetArgs migration.VolumeTargetArgs, op *operations.Operation) error {
```

Comment:
Same comment as CreateFromCopy, needs to change to match the fix I did on dir.

Reply:
Was actually mostly correct, suggesting this one got copy/pasted after I fixed `dir`.
Added logic to handle quota and changed reverter to match, should be good.